app/helpers/number_format_helper.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module NumberFormatHelper
+  def number_with_limit(number, opts = {})
+    init_formatting_options(opts) => { digits:, precision:, length_limit: }
+
+    string_number = number_with_precision(number, precision:, strip_insignificant_zeros: true)
+    length = string_number.delete("#{number_delimiter}#{number_separator}").length
+
+    scientific_notation_needed = length > length_limit ||
+                                 integer_part_size(number) > digits ||
+                                 (string_number == "0" && number != 0)
+
+    if scientific_notation_needed
+      format_scientific_notation(number, precision)
+    else
+      string_number
+    end
+  end
+
+  private
+
+  def integer_part_size(number) = number.round.to_s.size
+
+  def format_scientific_notation(number, precision)
+    str = "%0.*e" % [precision, number]
+
+    mantissa, exponent = str.split("e")
+
+    # normalize mantissa
+    mantissa = mantissa.sub(/(\.\d*?)0+$/, '\1')
```

Comment:
It's a bit hard to understand what the intention of this regular expression is. It might help to describe what exactly we are normalizing here.

Specifically I am wondering about this piece of the regex: `\d*?` The `?` seems superfluous, because `*` already matches zero repetitions of `\d`.

Reply:
`*?` is not superfluous, but a different matcher then `*` and `?`. It means, it expands lazy.

here is the difference between `/(\.\d+)0+$/` and `/(\.\d*?)0+$/`. Your suggestions grabs the zeros in a row greedy, as you already noticed above. This is not what we want to replace.

The second suggestion won't work for numbers that have only zeros after the comma, as those are not matches at all in the capturing group.